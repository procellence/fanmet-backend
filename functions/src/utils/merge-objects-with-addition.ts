/**
 * It will merge two objects and if there is common property,
 * it will add the value of the two properties in the output object
 * @param obj1
 * @param obj2
 */
export function mergeObjectsWithAddition(obj1: any, obj2: any): any {
  if (!obj1) return obj2 || null;
  if (!obj2) return obj1;

  const result = {} as any;

  const keys = new Set([...Object.keys(obj1), ...Object.keys(obj2)]);

  for (const key of keys) {
    if (typeof obj1[key] === 'object' && obj1[key] !== null && obj2[key]) {
      result[key] = mergeObjectsWithAddition(obj1[key], obj2[key]);
    } else if (key in obj1 && key in obj2 && typeof obj1[key] === typeof obj2[key]) {
      result[key] = (obj1[key]) + (obj2[key]);
    } else {
      result[key] = obj1[key] || obj2[key];
    }
  }

  return result;
}
