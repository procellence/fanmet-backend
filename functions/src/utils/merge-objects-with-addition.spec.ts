import { mergeObjectsWithAddition } from './merge-objects-with-addition';


describe('mergeObjectsWithAddition', () => {
  it('should handle two empty objects', () => {
    const obj1 = {};
    const obj2 = {};
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle one empty object', () => {
    const obj1 = { a: 1 };
    const obj2 = {};
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle objects with no overlapping keys', () => {
    const obj1 = { a: 1 };
    const obj2 = { b: 2 };
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle objects with overlapping keys but non-number values', () => {
    const obj1 = { a: 'hello' };
    const obj2 = { a: 'world' };
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle overlapping number keys', () => {
    const obj1 = { a: 1 };
    const obj2 = { a: 2 };
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle nested objects', () => {
    const obj1 = { a: { b: 1 } };
    const obj2 = { a: { b: 2, c: 3 } };
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });

  it('should handle mixed data types', () => {
    const obj1 = { a: 1, b: 'hello', c: { d: 3 } };
    const obj2 = { a: 2, b: 'world', c: { e: 4 } };
    const result = mergeObjectsWithAddition(obj1, obj2);
    expect(result).toMatchSnapshot();
  });
});
