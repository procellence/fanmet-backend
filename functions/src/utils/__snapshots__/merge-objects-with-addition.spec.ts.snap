// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeObjectsWithAddition should handle mixed data types 1`] = `
{
  "a": 3,
  "b": "helloworld",
  "c": {
    "d": 3,
    "e": 4,
  },
}
`;

exports[`mergeObjectsWithAddition should handle nested objects 1`] = `
{
  "a": {
    "b": 3,
    "c": 3,
  },
}
`;

exports[`mergeObjectsWithAddition should handle objects with no overlapping keys 1`] = `
{
  "a": 1,
  "b": 2,
}
`;

exports[`mergeObjectsWithAddition should handle objects with overlapping keys but non-number values 1`] = `
{
  "a": "helloworld",
}
`;

exports[`mergeObjectsWithAddition should handle one empty object 1`] = `
{
  "a": 1,
}
`;

exports[`mergeObjectsWithAddition should handle overlapping number keys 1`] = `
{
  "a": 3,
}
`;

exports[`mergeObjectsWithAddition should handle two empty objects 1`] = `{}`;
